rs/replicated_state/tests/system_state.rs

```diff
@@ -365,3 +397,109 @@ fn induct_messages_to_self_full_queue() {
+    let time = Time::from_nanos_since_unix_epoch(1);
+    let call_context_id = call_context_manager.new_call_context(
+        CallOrigin::SystemTask,
+        Cycles::zero(),
+        time,
+        RequestMetadata::new(0, time),
+    );
+
+    // Simulates an outbound call with the given deadline, by registering a callback
+    // and reserving a response slot.
+    let mut simulate_outbound_call = |deadline| {
+        // Register a callback.
+        let callback = Callback::new(
+            call_context_id,
+            CANISTER_ID,
+            OTHER_CANISTER_ID,
+            Cycles::zero(),
+            Cycles::new(42),
+            Cycles::new(84),
+            WasmClosure::new(0, 2),
+            WasmClosure::new(0, 2),
+            None,
+            deadline,
+        );
+        let call_context_manager = fixture.system_state.call_context_manager_mut().unwrap();
+        let callback_id = call_context_manager.register_callback(callback);
+
+        // Reserve a response slot.
+        fixture
+            .push_output_request(output_request(deadline))
+            .unwrap();
+        fixture.pop_output().unwrap();
+
+        callback_id
+    };
+
+    let deadline_expired_reject_payload = Payload::Reject(RejectContext::new(
+        RejectCode::SysUnknown,
+        "Call deadline has expired.",
+    ));
+
+    let d1 = CoarseTime::from_secs_since_unix_epoch(1);
+    let d2 = CoarseTime::from_secs_since_unix_epoch(2);
+    let d3 = CoarseTime::from_secs_since_unix_epoch(3);
+
+    let c1 = simulate_outbound_call(d1);
+    let c2 = simulate_outbound_call(d1);
+    let c3 = simulate_outbound_call(d1);
+    let c4 = simulate_outbound_call(d2);
+
+    // Simulate a paused execution for `c1`.
+    fixture
+        .push_input(input_response(c1, d1), InputQueueType::RemoteSubnet)
+        .unwrap();
+    let response1 = fixture.pop_input().unwrap();
+    fixture
+        .system_state
+        .task_queue
+        .push_front(ExecutionTask::PausedExecution {
+            id: PausedExecutionId(1),
+            input: CanisterMessageOrTask::Message(response1),
+        });
+
+    // And enqueue a response for `c2`.
+    fixture
+        .push_input(input_response(c2, d1), InputQueueType::RemoteSubnet)
+        .unwrap();
+
+    // Time out callbacks with deadlines before `d2` (only applicable to `c3` now).
+    assert_eq!(1, fixture.time_out_callbacks(d2));
+
+    // Pop the response for `c2`.
+    assert_matches!(
+        fixture.pop_input(),
+        Some(CanisterMessage::Response(response))
+            if response.originator_reply_callback == c2
+    );
+
+    // Pop the reject response for `c3`.
+    assert_matches!(
+        fixture.pop_input(),
+        Some(CanisterMessage::Response(response))
+            if response.originator_reply_callback == c3 && response.response_payload == deadline_expired_reject_payload
+    );
+    assert_eq!(None, fixture.pop_input());
+
```

Comment:
Should we peek the input here to see that before the timeout call the message is still untouched?

Reply:
Before the `time_out_callbacks()` call below, there are no messages, only the callback. That's what the `assert_eq!(None, pop_input())` above checks.

Reply:
Of course -- sorry, seems I was somehow confused. Thanks for clarifying.